File: visual-model-debugger/src/main/java/org/visual/model/debugger/core/DebuggerContext.java
package org.visual.model.debugger.core;

import io.avaje.inject.BeanScope;
import java.nio.charset.StandardCharsets;
import javafx.fxml.FXMLLoader;
import javafx.scene.Parent;
import lombok.Getter;
import lombok.SneakyThrows;
import lombok.val;
import org.jetbrains.annotations.NotNull;

@Getter
public enum DebuggerContext {

    INSTANCE;

    private final BeanScope beanScope = BeanScope.builder()
            .shutdownHook(true)
            .build();

    DebuggerContext(){
        System.err.println(beanScope.all());
    }

    public <T> @NotNull T get(Class<T> clazz) {
        return beanScope.get(clazz);
    }

    @SneakyThrows
    public Parent load(String prefix) {
        val loader = new FXMLLoader(VisualModelDebugger.class.getResource(prefix + ".fxml"));
        loader.setControllerFactory(beanScope::get);
        loader.setCharset(StandardCharsets.UTF_8);
        return loader.load();
    }
}

File: app/visual-model-debugger/src/main/kotlin/org/visual/model/debugger/context/DebuggerContext.kt
package org.visual.model.debugger.context

import io.avaje.inject.BeanScope
import java.nio.charset.StandardCharsets
import javafx.fxml.FXMLLoader
import javafx.scene.Parent
import javafx.util.Callback
import lombok.SneakyThrows
import org.visual.model.debugger.VisualModelDebugger

data object DebuggerContext {
  private val beanScope = BeanScope.builder().shutdownHook(true).build()

  fun <T> get(clazz: Class<T>): T {
    return beanScope.get(clazz)
  }

  @SneakyThrows
  fun load(prefix: String): Parent {
    val loader = FXMLLoader(VisualModelDebugger::class.java.getResource("$prefix.fxml"))
    loader.controllerFactory = Callback { aClass: Class<*> -> beanScope.get(aClass) }
    loader.charset = StandardCharsets.UTF_8
    return loader.load()
  }
}
